runtime/client_http_request.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zanzibar
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/uber-go/zap"
+)
+
+// ClientHTTPRequest is the struct for making outbound
+// requests using a client.
+type ClientHTTPRequest struct {
+	started     bool
+	startTime   time.Time
+	client      *HTTPClient
+	httpRequest *http.Request
+
+	ClientName string
+	MethodName string
+	Logger     zap.Logger
+}
+
+// NewClientHTTPRequest allocates a ClientHTTPRequest
+func NewClientHTTPRequest(
+	clientName string, methodName string,
+	client *HTTPClient,
+) *ClientHTTPRequest {
+	req := &ClientHTTPRequest{
+		Logger: client.Logger,
+		client: client,
+	}
+
+	req.start(clientName, methodName)
+	return req
+}
+
+// Start the request, do some metrics book keeping
+func (req *ClientHTTPRequest) start(
+	clientName string, methodName string,
+) {
+	if req.started {
+		/* coverage ignore next line */
+		req.Logger.Error(
+			"Cannot start ClientHTTPRequest twice",
+			zap.String("methodName", methodName),
+			zap.String("clientName", clientName),
+		)
+		/* coverage ignore next line */
+		return
+	}
+
+	req.ClientName = clientName
+	req.MethodName = methodName
+
+	req.started = true
+	req.startTime = time.Now()
+}
+
+// WriteJSON will send a json http request out.
+func (req *ClientHTTPRequest) WriteJSON(
```

Comment:
I'm looking to have symmetry with HTTPServerResponse class.

A client http request is definitely 1-1 with a single outbound http.Request.

I can rename this to `OutgoingHTTPRequest` instead of `ClientHTTPRequest` if that reduces confusion.